### ISSUE TITLE
Regex query params broken in 0.12.14+

### ISSUE THREAD

**Issue by royaldark** [angular-ui/ui-router#1998]
We use regex query params in conjunction with `$urlRouterProvider` to define a shortcut route for our users. This uses a simple regex matching one or more non-`/` characters. So if you access `ourserver.com/abc`, you should be redirected to `ourserver.com/other/route/abc/stuff`. However, if you access `ourserver.com/` that should not be matched and the `.otherwise` should bring you to `ourserver.com/home`. This works as expected in versions prior to 0.12.14, but beginning with that version, `ourserver.com/` matches the regex, even though the it requires `/[^\/]+/` See code sample below:

``` coffee
from = '/{paramA:[^/]+}'
to   = '/other/route/:paramA/stuff'

$urlRouterProvider.when from, ($match, $urlMatcherFactory) ->
  $urlMatcherFactory.compile(to).format($match)

$urlRouterProvider.otherwise '/home'
```
##### Version 0.12.13

Route: `/`
Expected: Redirect to `/home`
Actual: Redirect to `/home`
##### Version 0.12.14

Route: `/`
Expected: Redirect to `/home`
Actual: Redirect to `/other/route//stuff`

**Comment by christopherthielen** [comment]
Thanks, can you put this in a plunk?

**Comment by royaldark** [comment]
Here is a plunk showing correct behavior in 0.12.3: http://plnkr.co/edit/HrpbAPitgf1jRMDczCND?p=preview
And here is one showing the bad behavior in 0.12.5: http://plnkr.co/edit/14CND3QCaAySugrhiDHd?p=preview

They are exactly identical save for the differing UI Router version included.

Note that although the `/home` and `/12345` buttons work correctly in both versions, the `/` button redirects to `/home` in 0.12.3, but to `/content//view` in 0.12.5.

**Comment by christopherthielen** [comment]
just so it's clear, everywhere you reference 0.12.3 I assume you mean 0.2.13 and same for 0.12.5 meaning 0.2.15

**Comment by royaldark** [comment]
D'oh! Yes, thank you. I meant 0.2.13 and 0.2.15.

**Comment by ckniffen** [comment]
The main thing that changed that seems to be causing similar issues is the addition of $normalize.  Could that be causing the issue?

**Comment by jcreighton** [comment]
We're having a similar issue & it does appear to be due to $normalize. We are passing in a param with a `\` at the beginning: `{ path: '/article-title }`. In UrlMatcher.format, it is encoded 2x which results in `%252Farticle-title`. 

The problem being when you then refresh the page, the `%252Farticle-title` param is only decoded once which results in `%2Farticle-title`. What we need is it to be decoded 2x to result in `/article-title`, so it's in the proper format for the http request.

The $normalize function stops it from being encoded 2x as it's already been encoded by $location.

**Comment by NeishaChadwick** [comment]
+1 having almost the exact same circumstance as @royaldark. Upgraded from v0.2.11 to v0.2.15 and it stopped working. Our regex is `.[^/]+`, very similar. 

Our situation is that if you access `ourserver.com/things/abc`, you should be redirected to `ourserver.com/things/abc/overview`. However, if you access `ourserver.com/things/` that should not be matched and the defined state should handle it. Note that `ourserver.com/things` is a separate route altogether.

How we set up our old code:
```javascript
$urlRouterProvider.when('/thing/{thingName:.[^/]+}', 'thing/:thingName/overview');

$stateProvider
    .state('thing', {
        templateUrl: 'app/thing/thing.html',
        controller: 'ThingCtrl',
        params: { thingName: null }
    }).state('thing.overview', {
        url: '/things/:thingName/overview',
        templateUrl: 'app/thing/overview.html',
        controller: 'ThingOverviewCtrl'
    }).state('thing.new', {
        url: '/things/:thingName',
        templateUrl: 'app/thing/details.html',
        controller: 'ThingDetailsCtrl'
    });
```

Given this isn't working, here is our workaround:
```javascript
 $urlRouterProvider
    .when('/things/:thingName', ['$match', '$stateParams', function ($match, $stateParams) {
        // Checks if thingName exists, is not null and is not empty (all falsey)
        if (!!$match.thingName) {
            // If there is a thingName, navigate to the overview page for it 
            // From ui-router doco, can return "a String, which is treated as a redirect and passed to $location.url()"
            return '/things/' + $stateParams.thingName + '/overview';
        }
        // Returning false will then allow the route to be processed normally as defined in the $stateProvider config
        return false;
    }]);
The $stateProvider config stays the same...
```
It would be good to know if this is going to be fixed.

**Comment by royaldark** [comment]
Any progress on this? I don't want to stay on 0.12.13 forever, but this issue totally breaks our app.

We use regex query params in enough places that I don't want to have to copy-paste a workaround all over if I don't have to...

**Comment by linvi** [comment]
Same issue here, I think the issue come from the fact that the state internal object `url.regexp` is not correct.

If you try decorating the `state` like this (http://stackoverflow.com/questions/26824628/ui-router-state-current-wrapper-for-arbitary-state/26848546#26848546). The match will return an object with no parameter.

**Comment by christopherthielen** [comment]
Confirmed, there is still a bug in 0.2.17 here.

There is code that sets a default value of emptystring `""` for simple path parameters which don't specify a type (`string` is then chosen as the default type).  That code was attempting to provide backwards compatibility such that all path parameters are optional.  https://github.com/angular-ui/ui-router/blob/master/src/urlMatcherFactory.js#L923-L924

I have a note to remove this code when we get past 0.2.x (the feature-1.0 branch) because I think path parameters should be non-optional by default.  However this will be a significant breaking change for users.  Unfortunately, this logic is also kicking in for `.when()`


In the meantime, @royaldark I have a clean workaround for you: http://plnkr.co/edit/ASFecyma2Xbuz9a2mPcu?p=preview

You can make a custom type with a different name.  That custom type will avoid the "all path parameters are optional" logic.

```
  var stringType = $urlMatcherFactoryProvider.type("string");
  // Make a shallow copy of the string type, and override some values
  var requiredType = angular.extend({}, stringType);
  requiredType.name = 'required';
  requiredType.pattern = /[^/]+/;

  // Register custom 'required' type 
  $urlMatcherFactoryProvider.type("required", requiredType);
  ```
Now use it in your url redirect
```
  var from = '/{id:required}',
    to = '/content/:id/view';

  $urlRouterProvider.when(from, function($match, $urlMatcherFactory) {
    var newUrl = $urlMatcherFactory.compile(to).format($match);
    return newUrl;
  });
```

**Comment by christopherthielen** [comment]
@jcreighton  your issue is different.  It was fixed in 0.2.16 via 02e98660a80dfd1ca4b113dd24ee304af91e9f8c

**Comment by jcreighton** [comment]
@christopherthielen Thanks!
